<a href="https://colab.research.google.com/github/EduardaDalmas/python_bigO/blob/main/Teste_comparativo.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Trabalho Prático (10pt)


1. Agenda de Compromissos:

Crie um vetor ordenado para armazenar os compromissos, onde cada compromisso é um objeto com informações como data, hora, descrição, etc. 

a) Implemente um algoritmo para adicionar e excluir compromissos no vetor ordenado, mantendo sempre a ordem cronológica dos compromissos.

b) Implemente um algoritmo de busca binária para permitir aos usuários procurarem por compromissos específicos por data, hora ou outra informação relevante.

c) Implemente também a visualização de todos os compromissos em ordem cronológica.

# Código de Compromissos

In [34]:
import random

In [35]:
import datetime

In [36]:
class Compromisso:
  def __init__(self, data, hora, descricao):
    self.data = data
    self.hora = hora
    self.descricao = descricao

  def __lt__(self, other):
      return datetime.datetime.strptime(self.data + ' ' + self.hora, '%d/%m/%Y %H:%M') < datetime.datetime.strptime(other.data + ' ' + other.hora, '%d/%m/%Y %H:%M')

  def __str__(self):
    return f"{self.data} - {self.hora} - {self.descricao}"


class Agenda():
  def __init__(self, capacidade):
    self.capacidade = capacidade
    self.ultima_posicao = -1
    self.valores = [None] * self.capacidade

  def adicionar(self, valor):
    if self.ultima_posicao == self.capacidade -1:
      print('Capacidade máxima atingida')
      return 
    posicao = 0
    for i in range(self.ultima_posicao + 1):
      posicao = i
      if self.valores[i] > valor:
        break
      if i == self.ultima_posicao:
        posicao = i + 1
    x = self.ultima_posicao
    while x >= posicao:
      self.valores[x + 1] = self.valores[x]
      x -= 1
    self.valores[posicao] = valor
    self.ultima_posicao += 1 
    self.ordena_vetor(self.valores)

  def ordena_vetor(self, valores):
    valores_nao_vazios = filter(lambda valor: valor is not None, self.valores)
    ordenacao = sorted(valores_nao_vazios, key=lambda valor: valor.data, reverse=True)
    for i, valor in enumerate(ordenacao):
      print(f'indice {i} - Data: {valor.data} - hora: {valor.hora}, descricao: {valor.descricao}')

  def excluir(self, valor):
    if self.ultima_posicao == -1:
      print('Array vazio')
      return 
    else:
      for i in range(valor, self.ultima_posicao):
        self.valores[i] = self.valores[i + 1]
      self.ultima_posicao -= 1

  def buscar(self, valor, tipo):
    limite_inferior = 0
    limite_superior = self.ultima_posicao

    while True:
      posicao_atual = int((limite_inferior + limite_superior) / 2)
      
      if tipo == 'data':
        if self.valores[posicao_atual].data == valor:
          return posicao_atual

        elif limite_inferior > limite_superior:
          return -1
        
        else:
          if self.valores[posicao_atual].data < valor:
            limite_inferior = posicao_atual + 1
          else:
            limite_superior = posicao_atual - 1
      
      if tipo == 'hora':
        if self.valores[posicao_atual].hora == valor:
          return posicao_atual

        elif limite_inferior > limite_superior:
          return -1
        
        else:
          if self.valores[posicao_atual].hora < valor:
            limite_inferior = posicao_atual + 1
          else:
            limite_superior = posicao_atual - 1

      if tipo == 'descricao':
        if self.valores[posicao_atual].descricao == valor:
          return posicao_atual

        elif limite_inferior > limite_superior:
          return -1
        
        else:
          if self.valores[posicao_atual].descricao < valor:
            limite_inferior = posicao_atual + 1
          else:
            limite_superior = posicao_atual - 1      

  def visualizar(self):
    if self.ultima_posicao == -1:
      print('O vetor está vazio')   
    else:
      for i in range(self.ultima_posicao +1):
        compromisso = self.valores[i]
        print('posicao: ', i, ' | data: ', compromisso.data, ' | hora: ', compromisso.hora, ' | descricao: ', compromisso.descricao) 


*Inserir e listar*

In [29]:
agendar = Agenda(1000)

In [30]:
agendar.visualizar()

O vetor está vazio


*Adicionando registros*

In [ ]:
for i in range(1000):
  nomes = ['Reunião', 'Jantar', 'Call', 'Retrospectiva', 'Avaliação do time']
  tipo = ['de negocios', 'no Teams', 'no Meet', 'no Discord',]
  descricao_compromisso = random.choice(nomes) + " " + random.choice(tipo)
  ano = random.randint(1900, 2023)
  mes = random.randint(1, 12)
  if mes == 2:  # Fevereiro tem 28 ou 29 dias
    if ano % 4 == 0 and (ano % 100 != 0 or ano % 400 == 0):
        dia = random.randint(1, 29)
    else:
        dia = random.randint(1, 28)
  elif mes in (4, 6, 9, 11):  # Abril, Junho, Setembro e Novembro tem 30 dias
    dia = random.randint(1, 30)
  else:  # Os outros meses tem 31 dias
    dia = random.randint(1, 31)
  data_randomica = datetime.datetime(ano, mes, dia).strftime("%d/%m/%Y")
  hora = datetime.time(random.randint(0, 23), random.randint(0, 59)).strftime("%H:%M")
  novo_agendamento = Compromisso( 
    data_randomica,
    hora,
    descricao_compromisso)
  agendar.adicionar(novo_agendamento)

*Calculando tempo de inserção de 1000 registros*

In [37]:
elementos = []

for i in range(1000):
  nomes = ['Reunião', 'Jantar', 'Call', 'Retrospectiva', 'Avaliação do time']
  tipo = ['de negocios', 'no Teams', 'no Meet', 'no Discord',]
  descricao_compromisso = random.choice(nomes) + " " + random.choice(tipo)
  ano = random.randint(1900, 2023)
  mes = random.randint(1, 12)
  if mes == 2:  # Fevereiro tem 28 ou 29 dias
    if ano % 4 == 0 and (ano % 100 != 0 or ano % 400 == 0):
        dia = random.randint(1, 29)
    else:
        dia = random.randint(1, 28)
  elif mes in (4, 6, 9, 11):  # Abril, Junho, Setembro e Novembro tem 30 dias
    dia = random.randint(1, 30)
  else:  # Os outros meses tem 31 dias
    dia = random.randint(1, 31)
  data_randomica = datetime.datetime(ano, mes, dia).strftime("%d/%m/%Y")
  hora = datetime.time(random.randint(0, 23), random.randint(0, 59)).strftime("%H:%M")
  novo_agendamento = Compromisso( 
    data_randomica,
    hora,
    descricao_compromisso)
  elementos.append(novo_agendamento)

In [38]:
def insercao_agenda(elementos):
  for i in elementos:
    agendar.adicionar(i)
  return agendar

In [ ]:
insercao_agenda(elementos)

In [33]:
%timeit insercao_agenda(elementos)

A saída de streaming foi truncada nas últimas 5000 linhas.
Capacidade máxima atingida
Capacidade máxima atingida
Capacidade máxima atingida
Capacidade máxima atingida
Capacidade máxima atingida
Capacidade máxima atingida
Capacidade máxima atingida
Capacidade máxima atingida
Capacidade máxima atingida
Capacidade máxima atingida
Capacidade máxima atingida
Capacidade máxima atingida
Capacidade máxima atingida
Capacidade máxima atingida
Capacidade máxima atingida
Capacidade máxima atingida
Capacidade máxima atingida
Capacidade máxima atingida
Capacidade máxima atingida
Capacidade máxima atingida
Capacidade máxima atingida
Capacidade máxima atingida
Capacidade máxima atingida
Capacidade máxima atingida
Capacidade máxima atingida
Capacidade máxima atingida
Capacidade máxima atingida
Capacidade máxima atingida
Capacidade máxima atingida
Capacidade máxima atingida
Capacidade máxima atingida
Capacidade máxima atingida
Capacidade máxima atingida
Capacidade máxima atingida
Capacidade máxima ating

*Visualizando os registros inseridos*

In [42]:
agendar.visualizar()

posicao:  0  | data:  01/01/2018  | hora:  02:47  | descricao:  Jantar no Teams
posicao:  1  | data:  01/02/1921  | hora:  18:17  | descricao:  Jantar no Meet
posicao:  2  | data:  01/02/1962  | hora:  04:11  | descricao:  Jantar de negocios
posicao:  3  | data:  01/02/1998  | hora:  09:32  | descricao:  Reunião de negocios
posicao:  4  | data:  01/02/2016  | hora:  06:10  | descricao:  Jantar no Teams
posicao:  5  | data:  01/03/1914  | hora:  08:39  | descricao:  Reunião no Teams
posicao:  6  | data:  01/03/1962  | hora:  01:39  | descricao:  Avaliação do time no Teams
posicao:  7  | data:  01/03/2005  | hora:  06:19  | descricao:  Call no Discord
posicao:  8  | data:  01/04/1914  | hora:  00:27  | descricao:  Jantar de negocios
posicao:  9  | data:  01/04/1943  | hora:  00:32  | descricao:  Retrospectiva no Meet
posicao:  10  | data:  01/06/1963  | hora:  17:40  | descricao:  Reunião no Discord
posicao:  11  | data:  01/06/1978  | hora:  02:26  | descricao:  Avaliação do time de neg

*Excluir*

In [40]:
agendar.excluir(1)

*Buscar*

In [43]:
agendar.buscar(valor='Retrospectiva no Teams', tipo='descricao')

499

In [44]:
agendar.buscar(valor='16/05/1922', tipo='data')

496

*Busca por 1000 registros*

In [45]:
def busca_agenda(elementos):
  for i in elementos:
    agendar.buscar(valor=i.descricao, tipo='descricao')
  return agendar


In [46]:
%timeit busca_agenda(elementos)

6.34 ms ± 2.03 ms per loop (mean ± std. dev. of 7 runs, 100 loops each)


2. Estoque de uma Loja Online:

Crie um catálogo de produtos em um vetor não ordenado, onde cada elemento contém informações sobre o produto, como nome, código, quantidade, preço e avaliação de clientes. 

a) Implemente um algoritmo para adicionar e excluir produtos no vetor não ordenado.

b) Implemente  um algoritmo de busca para permitir aos funcionários procurarem por produtos específicos por código, nome ou outra informação relevante.

c) Implemente um algoritmo de busca ou ordenação para encontrar os melhores produtos, por exemplo, aqueles com as melhores avaliações ou melhores preços. Retorne os produtos encontrados para exibição aos usuários.

d) Implemente também a visualização de todos os produtos do estoque e cálculo de total de estoque.

In [64]:
class Produto:
  def __init__(self, nome, codigo, quantidade, preco, avaliacao):
    self.nome = nome
    self.codigo = codigo
    self.quantidade = quantidade
    self.preco = preco
    self.avaliacao = avaliacao

class Estoque():
  def __init__(self, capacidade):
    self.capacidade = capacidade
    self.ultima_posicao = -1
    self.produtos = [None] * self.capacidade

  def adicionar(self, valor):
    if self.ultima_posicao == self.capacidade -1:
      print('Capacidade máxima atingida')
    else:
      self.ultima_posicao  += 1
      self.produtos[self.ultima_posicao] = valor
  
  def buscar(self, valor, tipo):
    if tipo == 'codigo':
      for i in range(self.ultima_posicao + 1):
        codigo_produto = self.produtos[i]
        if int(valor) == int(codigo_produto.codigo):
          return i
      return -1

    if tipo == 'nome':
      for i in range(self.ultima_posicao + 1):
        codigo_produto = self.produtos[i]
        if valor == codigo_produto.nome:
          return i
      return -1
  
  def ordenar(self, tipo):
    if tipo == 'preco':
      produtos_nao_vazios = filter(lambda produto: produto is not None, self.produtos)
      ordenacao = sorted(produtos_nao_vazios, key=lambda produto: produto.preco, reverse=True)
      for i, produto in enumerate(ordenacao):
        print(f'indice {i} - Nome: {produto.nome} - código: {produto.codigo}, quantidade: {produto.quantidade}, preço: {produto.preco}, avaliações: {produto.avaliacao}')
    
    if tipo == 'avaliacao':
      produtos_nao_vazios = filter(lambda produto: produto is not None, self.produtos)
      ordenacao = sorted(produtos_nao_vazios, key=lambda produto: produto.avaliacao, reverse=True)
      for i, produto in enumerate(ordenacao):
        print(f'indice {i} - Nome: {produto.nome} - código: {produto.codigo}, quantidade: {produto.quantidade}, preço: {produto.preco}, avaliações: {produto.avaliacao}')

  def excluir(self, valor, tipo):
    posicao = self.buscar(valor, tipo)
    if posicao == -1:
       return -1
    else:
      for i in range(posicao, self.ultima_posicao):
        self.produtos[i] = self.produtos[i + 1]
      self.ultima_posicao -= 1

  def visualizar(self):
    if self.ultima_posicao == -1:
      print('O vetor está vazio')   
    else:
      for i in range(self.ultima_posicao +1):
        produto = self.produtos[i]
        print('posicao: ', i, ' | nome: ', produto.nome, ' | codigo: ', produto.codigo, ' | quantidade: ', produto.quantidade, ' | preco: ', produto.preco, ' | avaliacao: ', produto.avaliacao) 
        
  def calcular(self): 
    for i in range(self.ultima_posicao +1):
      produto = self.produtos[i]
      total = int(produto.quantidade) * int(produto.preco)
    return total

*Insere e lista*

In [65]:
estocar = Estoque(1000)

In [66]:
estocar.visualizar()

O vetor está vazio


*Inserir 10 registros para teste*



In [ ]:
for i in range(10):
  nomes = ['Livro', 'Caderno', 'Caneta', 'Borracha', 'Lapis']
  marcas = ['Big', 'Fabber Castel', 'Foroni', 'Disney', 'Marvel']
  nome_produto = random.choice(nomes) + " " + random.choice(marcas)
  novo_produto = Produto(
    nome_produto,
    str(random.randint(1, 1000)),
    str(random.randint(1, 100)),
    str(random.randint(1, 10)),
    str(random.randint(1, 5)))
  estocar.adicionar(novo_produto)

*Inserir 1000 registros*

In [67]:
elementos = []

for i in range(1000):
  nomes = ['Livro', 'Caderno', 'Caneta', 'Borracha', 'Lapis']
  marcas = ['Big', 'Fabber Castel', 'Foroni', 'Disney', 'Marvel']
  nome_produto = random.choice(nomes) + " " + random.choice(marcas)
  novo_produto = Produto(
    nome_produto,
    str(random.randint(1, 1000)),
    str(random.randint(1, 100)),
    str(random.randint(1, 10)),
    str(random.randint(1, 5)))
  elementos.append(novo_produto)

In [68]:
def insere_produto(elementos):
  for i in elementos:
    estocar.adicionar(i)
  return agendar


In [69]:
%timeit insere_produto(elementos)

A saída de streaming foi truncada nas últimas 5000 linhas.
Capacidade máxima atingida
Capacidade máxima atingida
Capacidade máxima atingida
Capacidade máxima atingida
Capacidade máxima atingida
Capacidade máxima atingida
Capacidade máxima atingida
Capacidade máxima atingida
Capacidade máxima atingida
Capacidade máxima atingida
Capacidade máxima atingida
Capacidade máxima atingida
Capacidade máxima atingida
Capacidade máxima atingida
Capacidade máxima atingida
Capacidade máxima atingida
Capacidade máxima atingida
Capacidade máxima atingida
Capacidade máxima atingida
Capacidade máxima atingida
Capacidade máxima atingida
Capacidade máxima atingida
Capacidade máxima atingida
Capacidade máxima atingida
Capacidade máxima atingida
Capacidade máxima atingida
Capacidade máxima atingida
Capacidade máxima atingida
Capacidade máxima atingida
Capacidade máxima atingida
Capacidade máxima atingida
Capacidade máxima atingida
Capacidade máxima atingida
Capacidade máxima atingida
Capacidade máxima ating

*Buscar por 1000 registros*

In [70]:
def busca_produto(elementos):
  for i in elementos:
    estocar.buscar(i.codigo, 'codigo')
  return agendar

In [71]:
%timeit busca_produto(elementos)

138 ms ± 36.5 ms per loop (mean ± std. dev. of 7 runs, 10 loops each)


*Ordenar*

In [72]:
estocar.ordenar('preco')

indice 0 - Nome: Borracha Marvel - código: 88, quantidade: 59, preço: 9, avaliações: 3
indice 1 - Nome: Lapis Marvel - código: 220, quantidade: 41, preço: 9, avaliações: 1
indice 2 - Nome: Caderno Foroni - código: 948, quantidade: 48, preço: 9, avaliações: 3
indice 3 - Nome: Livro Disney - código: 936, quantidade: 2, preço: 9, avaliações: 1
indice 4 - Nome: Livro Disney - código: 785, quantidade: 19, preço: 9, avaliações: 5
indice 5 - Nome: Caderno Disney - código: 945, quantidade: 52, preço: 9, avaliações: 2
indice 6 - Nome: Caderno Foroni - código: 369, quantidade: 28, preço: 9, avaliações: 3
indice 7 - Nome: Livro Marvel - código: 648, quantidade: 59, preço: 9, avaliações: 2
indice 8 - Nome: Livro Fabber Castel - código: 971, quantidade: 43, preço: 9, avaliações: 5
indice 9 - Nome: Livro Big - código: 641, quantidade: 29, preço: 9, avaliações: 5
indice 10 - Nome: Lapis Fabber Castel - código: 540, quantidade: 9, preço: 9, avaliações: 2
indice 11 - Nome: Livro Big - código: 667, quan

In [73]:
estocar.ordenar('avaliacao')

indice 0 - Nome: Caderno Disney - código: 17, quantidade: 22, preço: 7, avaliações: 5
indice 1 - Nome: Caneta Big - código: 836, quantidade: 94, preço: 10, avaliações: 5
indice 2 - Nome: Lapis Disney - código: 934, quantidade: 77, preço: 1, avaliações: 5
indice 3 - Nome: Caneta Marvel - código: 430, quantidade: 88, preço: 10, avaliações: 5
indice 4 - Nome: Borracha Disney - código: 931, quantidade: 98, preço: 1, avaliações: 5
indice 5 - Nome: Caderno Fabber Castel - código: 216, quantidade: 9, preço: 4, avaliações: 5
indice 6 - Nome: Borracha Marvel - código: 156, quantidade: 94, preço: 7, avaliações: 5
indice 7 - Nome: Caderno Marvel - código: 294, quantidade: 98, preço: 1, avaliações: 5
indice 8 - Nome: Caderno Big - código: 506, quantidade: 51, preço: 3, avaliações: 5
indice 9 - Nome: Livro Foroni - código: 660, quantidade: 52, preço: 6, avaliações: 5
indice 10 - Nome: Livro Big - código: 266, quantidade: 86, preço: 6, avaliações: 5
indice 11 - Nome: Caneta Disney - código: 794, qua

In [74]:
estocar.calcular()

1000


584

*Buscar*

In [75]:
estocar.buscar("Livro Marvel", 'nome')

12

*Excluir*

In [76]:
estocar.excluir(112, 'codigo')

In [77]:
estocar.visualizar()

posicao:  0  | nome:  Caderno Disney  | codigo:  17  | quantidade:  22  | preco:  7  | avaliacao:  5
posicao:  1  | nome:  Caneta Big  | codigo:  836  | quantidade:  94  | preco:  10  | avaliacao:  5
posicao:  2  | nome:  Lapis Disney  | codigo:  934  | quantidade:  77  | preco:  1  | avaliacao:  5
posicao:  3  | nome:  Caneta Marvel  | codigo:  430  | quantidade:  88  | preco:  10  | avaliacao:  5
posicao:  4  | nome:  Lapis Foroni  | codigo:  805  | quantidade:  58  | preco:  4  | avaliacao:  2
posicao:  5  | nome:  Lapis Fabber Castel  | codigo:  284  | quantidade:  15  | preco:  5  | avaliacao:  3
posicao:  6  | nome:  Caneta Fabber Castel  | codigo:  707  | quantidade:  40  | preco:  5  | avaliacao:  2
posicao:  7  | nome:  Caneta Disney  | codigo:  258  | quantidade:  61  | preco:  8  | avaliacao:  2
posicao:  8  | nome:  Lapis Marvel  | codigo:  771  | quantidade:  28  | preco:  10  | avaliacao:  3
posicao:  9  | nome:  Caneta Disney  | codigo:  531  | quantidade:  15  | preco: 

4) Responda:

**a) Qual foi o melhor tempo de execução de adição e de qual sistema?**
Inserção no sistema de Produtos, com 288 ms


**b) Qual foi o melhor tempo de execução de busca e de qual sistema?**
Busca Binária no sistema de Agenda, com 6.95 ms


Agenda inserção: 315 ms O(n)

Agenda Busca: 6.95 ms O(log n)

Produto inserção: 288 ms O(1)

Produto Busca: 138 ms O(n)